test/e2e/user_certs_test.go

```diff
@@ -0,0 +1,162 @@
+	"crypto/x509"
+	"net/url"
+	"strings"
+	"testing"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientcorev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/util/cert"
+
+	configv1 "github.com/openshift/api/config/v1"
+	configclient "github.com/openshift/client-go/config/clientset/versioned/typed/config/v1"
+	operatorclient "github.com/openshift/client-go/operator/clientset/versioned/typed/operator/v1"
+
+	test "github.com/openshift/cluster-kube-apiserver-operator/test/library"
+)
+
+func TestNamedCertificates(t *testing.T) {
+
+	kubeConfig, err := test.NewClientConfigForTest()
+	test.FatalIfErr(t, err)
+	kubeClient, err := clientcorev1.NewForConfig(kubeConfig)
+	test.FatalIfErr(t, err)
+	configClient, err := configclient.NewForConfig(kubeConfig)
+	test.FatalIfErr(t, err)
+	operatorClient, err := operatorclient.NewForConfig(kubeConfig)
+	test.FatalIfErr(t, err)
+
+	initialConfigGeneration := test.GetKubeAPIServerOperatorConfigGeneration(t, operatorClient)
+
+	t.Log("Given the kube-apiserver-operator is available.")
+	test.WaitForKubeAPIServerClusterOperatorAvailableNotProgressingNotFailing(t, configClient)
+
+	t.Log("And some new tls Secrets have been created in the openshift-config namespace.")
+	rootCA := test.NewCertificateAuthorityCertificate(t, nil)
+
+	type testCase struct {
+		serverName           string
+		expectedSerialNumber string
+	}
+	var testCases []testCase
+
+	var namedCertificates []configv1.APIServerNamedServingCert
+
+	// create servingCerts and TLS Secrets, building testcases and name certificates config along the way
+	namedServingCertHosts := [][]string{
+		{"one.test"},
+		{"two.test"},
+		{"three.test", "four.test"},
+	}
+	for _, hosts := range namedServingCertHosts {
+		tls := test.NewServerCertificate(t, rootCA.Certificate, hosts...)
+		secretName := strings.ToLower(test.GenerateNameForTest(t, "Server"))
+		defer func() {
+			if err := kubeClient.Secrets("openshift-config").Delete(secretName, &metav1.DeleteOptions{}); err != nil {
+				t.Log(err)
+			}
+		}()
+		_, err := kubeClient.Secrets("openshift-config").Create(
+			&corev1.Secret{
+				ObjectMeta: metav1.ObjectMeta{Name: secretName},
+				Type:       corev1.SecretTypeTLS,
+				Data: map[string][]byte{
+					corev1.TLSPrivateKeyKey: cert.EncodePrivateKeyPEM(tls.PrivateKey),
+					corev1.TLSCertKey:       cert.EncodeCertPEM(tls.Certificate),
+				},
+			})
+		test.FatalIfErr(t, err)
+		namedCertificates = append(namedCertificates, configv1.APIServerNamedServingCert{ServingCertificate: configv1.SecretNameReference{Name: secretName}})
+		for _, host := range hosts {
+			testCases = append(testCases, testCase{
+				serverName:           host,
+				expectedSerialNumber: tls.Certificate.SerialNumber.String(),
+			})
+		}
+	}
+
+	t.Log("When the cluster is configured the use of the named certificates.")
+	defer func() {
+		t.Log("Restoring global cluster configuration")
```

Comment:
wondering, everything we do here in the e2e test, are there side-effect to other tests? Are there other tests running at all? <--- am just curious. Don't know the e2e context for our operators.